File: Anexos_Semana_01/Anexo 01.5-Ejemplos OpenCV en C++/ejemplo3.cpp
// CC5213 - Recuperación de Información Multimedia
// Profesor: Juan Manuel Barrios
// Fecha: 10 de marzo de 2025

#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/objdetect.hpp>

int main(int argc, char **argv) {
	if (argc < 5) {
		std::cout << "Uso:" << std::endl;
		std::cout << " " << argv[0] << std::endl;
		std::cout << "   id_webcam      Webcam. Ej: 0, 1, 2 " << std::endl;
		std::cout << "   XML_detector   Archivo detector. Ej: haarcascade_frontalface_default.xml" << std::endl;
		std::cout << "   scale_factor   Reducción de la imagen. 1.01 (1%=lento), 1.1 (10%=normal), 1.5 (50%=rápido)"
				<< std::endl;
		std::cout << "   min_neighbors  Cuantas veces se debe detectar. 1 (ruidoso), 5 (confiable), 20 (seguro)"
				<< std::endl;
		return 1;
	}

	//parametros de entrada
	int id_webcam = std::atoi(argv[1]);
	std::string xml = argv[2];
	double scale_factor = std::atof(argv[3]);
	int min_neighbors = std::atoi(argv[4]);

	//cargar el detector
	std::string file_xml = cv::samples::findFile(xml);
	std::cout << "cargando " << file_xml << std::endl;
	cv::CascadeClassifier detector;
	if (!detector.load(xml)) {
		std::cout << "error leyendo " << xml << std::endl;
		return 1;
	}
	std::cout << "OK " << xml << std::endl;

	//abrir webcam
	std::cout << "abriendo webcam " << id_webcam << std::endl;
	cv::VideoCapture capture;
	capture.open(id_webcam);
	if (!capture.isOpened()) {
		std::cout << "error abriendo " << id_webcam << std::endl;
		return 1;
	}
	std::cout << "OK webcam " << id_webcam << std::endl;

	//procesar frames
	std::string window_name = "webcam " + std::to_string(id_webcam);
	cv::namedWindow(window_name, cv::WINDOW_AUTOSIZE);
	cv::Mat frame_color, frame_gris;
	while (capture.grab()) {
		if (!capture.retrieve(frame_color))
			break;
		cv::cvtColor(frame_color, frame_gris, cv::COLOR_BGR2GRAY);
		std::vector<cv::Rect> detecciones;
		detector.detectMultiScale(frame_gris, detecciones, scale_factor, min_neighbors);
		std::cout << detecciones.size() << " detecciones" << std::endl;
		for (const cv::Rect &deteccion : detecciones) {
			cv::Point center(deteccion.x + deteccion.width / 2, deteccion.y + deteccion.height / 2);
			cv::Size size(deteccion.width / 2, deteccion.height / 2);
			cv::Scalar color(255, 255, 0);
			int thickness = 3;
			cv::ellipse(frame_color, center, size, 0, 0, 360, color, thickness);
		}
		cv::imshow(window_name, frame_color);
		char c = cv::waitKey(1);
		if (c == 27 || c == 'q') //tecla ESC o q
			break;
	}
	return 0;
}

File: Anexos_Semana_01/Anexo 01.4-Ejemplos OpenCV en python/ejemplo3.py
# CC5213 - Recuperación de Información Multimedia
# Profesor: Juan Manuel Barrios
# Fecha: 10 de marzo de 2025

import sys
import os
import cv2


def open_video(filename):
    capture = None
    if filename.isdigit():
        capture = cv2.VideoCapture(int(filename), cv2.CAP_DSHOW)
    elif os.path.isfile(filename):
        capture = cv2.VideoCapture(filename)
    if capture is None or not capture.isOpened():
        print("error abriendo {}".format(filename))
        sys.exit()
    return capture


def open_detector(xml):
    file_xml = cv2.data.haarcascades + xml
    print("leyendo {}".format(file_xml))
    detector = cv2.CascadeClassifier()
    if not detector.load(file_xml):
        print("no puedo cargar {}".format(xml))
        sys.exit()
    return detector


def ejemplo3(filename, xml, scale_factor, min_neighbors):
    # cargar el detector
    detector = open_detector(xml)
    # abrir video
    capture = open_video(filename)
    # procesar frames
    window_name = "video " + filename
    cv2.namedWindow(window_name, cv2.WINDOW_AUTOSIZE)
    while capture.grab():
        retval, frame_color = capture.retrieve()
        if not retval:
            continue
        frame_gris = cv2.cvtColor(frame_color, cv2.COLOR_BGR2GRAY)
        detecciones = detector.detectMultiScale(frame_gris, scale_factor, min_neighbors)
        print("detecciones: {}".format(len(detecciones)))
        for (x, y, w, h) in detecciones:
            center = (int(x + w / 2), int(y + h / 2))
            size = (int(w / 2), int(h / 2))
            color = (255, 255, 0)
            thickness = 3
            cv2.ellipse(frame_color, center, size, 0, 0, 360, color, thickness)
        cv2.imshow(window_name, frame_color)
        key = cv2.waitKey(1)
        if key == ord('q') or key == 27:
            break
    capture.release()
    cv2.destroyAllWindows()


if len(sys.argv) < 5:
    print("Uso:")
    print(" {}".format(sys.argv[0]))
    print("   video          Video o Id de la webcam. Ej: 0, 1, 2 ")
    print("   XML_detector   Archivo detector. Ej: haarcascade_frontalface_default.xml")
    print("   scale_factor   Reducción de la imagen. 1.01 (1%=lento), 1.1 (10%=normal), 1.5 (50%=rápido)")
    print("   min_neighbors  Cuantas veces se debe detectar. 1 (ruidoso), 5 (confiable), 20 (seguro)")
    sys.exit()

# parametros de entrada
input_filename = sys.argv[1]
input_xml = sys.argv[2]
input_scale_factor = float(sys.argv[3])
input_min_neighbors = int(sys.argv[4])

ejemplo3(input_filename, input_xml, input_scale_factor, input_min_neighbors)
